File: clients/cli/cmd/bamboo/relay_picker.go
// SPDX-License-Identifier: Apache-2.0

package main

import (
	"context"
	"fmt"
	"net"
	"sort"
	"sync"
	"time"

	bamboov1 "github.com/hanfour/bamboo/proto/gen/go/bamboo/v1"
)

// relayProbeTimeout caps each per-relay TCP-connect probe. Tighter
// than the controller's 3s /healthz probe because here we're racing
// to bring the tunnel up — every second spent probing is a second
// the user sees nothing. 1s is enough to fingerprint inter-region
// latency (typical cross-region is 50-300ms, intra-region <30ms);
// anything slower is already too slow to be the preferred relay.
var relayProbeTimeout = 1 * time.Second

// relayRegionToleranceMs is the §4 P2 stage 5 affinity window. A
// same-region relay wins over a lower-RTT out-of-region relay when
// the RTT gap is within this many milliseconds — the same magnitude
// as typical intra-AZ jitter, so we're not preferring slowness for
// the sake of geography. Beyond the tolerance the cross-region
// option is genuinely faster and wins.
//
// 50ms picked from observation: trans-continental RTT (say
// us-east-1 ↔ ap-northeast-1) is 150-200ms; intra-region jitter is
// typically <20ms. A 50ms tolerance absorbs the jitter without ever
// letting a far-away relay win when it shouldn't.
var relayRegionToleranceMs = int64(50)

// probeFn is the seam tests use to swap in a deterministic latency
// table for pickRelay's ranking logic. Production binds it to
// probeRelayRTT at init; tests overwrite with a closure that maps
// hostname → simulated RTT. Kept as a package var rather than a
// parameter on pickRelay to keep the call site in maybeOpenRelay
// quiet — only tests care about the seam.
var probeFn = probeRelayRTT

// rankedRelay is an internal record of (candidate index, measured
// RTT) used to rank probes. Hoisted to file scope so the region-
// affinity helper can take a slice of it without re-declaring an
// anonymous structurally-equivalent type that wouldn't be
// assignable across function boundaries.
type rankedRelay struct {
	idx int
	rtt time.Duration
}

// pickRelay selects the lowest-RTT relay from a controller-supplied
// list (§4 P2 multi-relay stage 2). RTT is measured as TCP connect
// time to (hostname, port) — same 5-tuple a real relay session
// would use, so a stuck firewall on the relay port surfaces here
// too.
//
// preferredRegion is the §4 P2 stage 5 affinity hint. Empty ⇒ pure
// RTT ranking (back-compatible with pre-stage-5 behaviour). When
// set, a same-region relay wins over a lower-RTT out-of-region
// relay if the RTT gap is within relayRegionToleranceMs. The point
// is to stop the picker flapping between two regions whose RTT
// numbers happen to land close — pick a "home" and stay there.
//
// Returns nil + nil when the list is empty (caller falls back to
// the env-var override path). Returns nil + the joined error when
// every probe fails — the caller treats that as "no usable relay"
// and continues without one rather than blocking tunnel bring-up.
//
// Parallel probes by design: serial would stretch a 5-relay probe
// to 5s in the worst case, doubling the user's perceived bring-up
// latency. The contention overhead at this fan-out is negligible.
func pickRelay(ctx context.Context, candidates []*bamboov1.RelayServer, preferredRegion string) (*bamboov1.RelayServer, error) {
	if len(candidates) == 0 {
		return nil, nil
	}
	type result struct {
		idx int
		rtt time.Duration
		err error
	}
	results := make(chan result, len(candidates))
	var wg sync.WaitGroup
	wg.Add(len(candidates))
	for i, rs := range candidates {
		go func(idx int, hostname string, port int32) {
			defer wg.Done()
			rtt, err := probeFn(ctx, hostname, int(port))
			results <- result{idx: idx, rtt: rtt, err: err}
		}(i, rs.GetHostname(), rs.GetPort())
	}
	wg.Wait()
	close(results)

	var healthy []rankedRelay
	var errs []error
	for r := range results {
		if r.err != nil {
			errs = append(errs, fmt.Errorf("%s: %w", candidates[r.idx].GetHostname(), r.err))
			continue
		}
		healthy = append(healthy, rankedRelay{idx: r.idx, rtt: r.rtt})
	}
	if len(healthy) == 0 {
		return nil, joinErrors(errs)
	}
	sort.SliceStable(healthy, func(a, b int) bool { return healthy[a].rtt < healthy[b].rtt })
	winnerIdx := applyRegionAffinity(candidates, healthy[0].idx, healthy[0].rtt, healthy, preferredRegion)
	return candidates[winnerIdx], nil
}

// applyRegionAffinity returns the index of the winning relay after
// the §4 P2 stage 5 same-region preference. preferredRegion empty
// ⇒ the RTT winner stays. Otherwise: if the RTT winner already is
// in the preferred region, no-op. If a same-region relay exists
// within toleranceMs of the RTT winner's latency, prefer it
// (lowest-RTT among same-region rows wins the tiebreaker — sort
// order from the caller is already ASC, so the first match wins).
//
// Returns the original winnerIdx when no eligible same-region
// candidate is found, so a region hint can never make the picker
// worse — only same or better, never slower than tolerance.
func applyRegionAffinity(
	candidates []*bamboov1.RelayServer,
	winnerIdx int,
	winnerRTT time.Duration,
	healthy []rankedRelay,
	preferredRegion string,
) int {
	if preferredRegion == "" {
		return winnerIdx
	}
	if candidates[winnerIdx].GetRegion() == preferredRegion {
		return winnerIdx
	}
	tolerance := time.Duration(relayRegionToleranceMs) * time.Millisecond
	for _, r := range healthy {
		if candidates[r.idx].GetRegion() != preferredRegion {
			continue
		}
		if r.rtt-winnerRTT <= tolerance {
			return r.idx
		}
		// healthy is ASC by RTT; once a same-region row exceeds
		// the tolerance, all later ones do too. Break.
		break
	}
	return winnerIdx
}

// probeRelayRTT measures TCP-connect latency to one relay endpoint.
// Closed immediately after measurement — we only care about the
// dial; reading bytes would add a TLS handshake's worth of variance
// that doesn't reflect what matters (will my WG-over-WSS session
// land here quickly).
//
// Returns the measured RTT on success. On failure (timeout, refused,
// DNS fail) returns 0 + the underlying error so the caller can log
// what's wrong with which relay.
func probeRelayRTT(ctx context.Context, hostname string, port int) (time.Duration, error) {
	probeCtx, cancel := context.WithTimeout(ctx, relayProbeTimeout)
	defer cancel()
	dialer := &net.Dialer{}
	start := time.Now()
	conn, err := dialer.DialContext(probeCtx, "tcp", fmt.Sprintf("%s:%d", hostname, port))
	rtt := time.Since(start)
	if err != nil {
		return 0, err
	}
	_ = conn.Close()
	return rtt, nil
}

// relayWSSURL builds the wss URL for a chosen relay. The controller's
// shape stores (hostname, port) — the /relay path + wss:// scheme
// are convention, not data. Pulled out so a future change to the
// scheme (e.g. h3) lives in one place instead of being threaded
// through every dial site.
func relayWSSURL(rs *bamboov1.RelayServer) string {
	return fmt.Sprintf("wss://%s:%d/relay", rs.GetHostname(), rs.GetPort())
}

// joinErrors collapses a slice of probe errors into one error
// suitable for logging. Returns nil when the slice is empty.
// Single-error case is unwrapped so the caller sees the bare cause
// rather than a "[...]" wrapper.
func joinErrors(errs []error) error {
	if len(errs) == 0 {
		return nil
	}
	if len(errs) == 1 {
		return errs[0]
	}
	msg := "all relay probes failed:"
	for _, e := range errs {
		msg += " [" + e.Error() + "]"
	}
	return fmt.Errorf("%s", msg)
}

File: clients/apple/Shared/RelayPicker.swift
// SPDX-License-Identifier: Apache-2.0

import Foundation
import Network

/// RelayPicker selects the lowest-RTT relay from a controller-
/// supplied list (§4 P2 multi-relay stage 2 — Apple parity to
/// PR #193's CLI implementation).
///
/// Selection mechanics mirror the CLI: parallel TCP-connect
/// probes via NWConnection, pick the lowest-RTT eligible row.
/// Picking on connect time (not HTTP /healthz round-trip)
/// matches the 5-tuple a real relay session uses, so a stuck
/// firewall on the relay port surfaces here rather than later
/// as a handshake failure.
///
/// The user-typed Settings `relayURL` is honored as an override
/// upstream in ConnectionViewModel — RelayPicker only runs when
/// that field is empty.
internal enum RelayPicker {
    /// probeTimeout caps each per-relay probe. Tighter than the
    /// controller's 3s /healthz cadence because the Apple client
    /// is racing to bring the tunnel up; every second spent
    /// probing is a second the user sees a spinner. 1s fingerprints
    /// inter-region latency cleanly (typical 50-300ms cross-region,
    /// <30ms intra-region); anything slower would be too slow to
    /// prefer anyway.
    static let probeTimeout: TimeInterval = 1.0

    /// ProbeFn is the test seam — production binds it to
    /// probeRelayRTT (real TCP connect via NWConnection); tests
    /// inject a deterministic latency table. Latency-driven
    /// ranking can't be tested with localhost probes (the kernel
    /// completes the SYN-ACK before any user-space callback runs),
    /// so isolating the seam is the only way to lock the ASC sort
    /// direction without flakiness.
    typealias ProbeFn = (_ hostname: String, _ port: Int) async -> Result<TimeInterval, Error>

    static var probeFn: ProbeFn = probeRelayRTT

    /// regionToleranceSeconds is the §4 P2 stage 5 affinity window.
    /// A same-region relay wins over a lower-RTT out-of-region one
    /// when the RTT gap is within this many seconds — same magnitude
    /// as typical intra-AZ jitter, so the picker isn't preferring
    /// slowness for the sake of geography. Beyond the tolerance the
    /// cross-region option is genuinely faster and wins.
    ///
    /// 50ms (0.05s) matches the CLI's relayRegionToleranceMs so both
    /// clients pick the same relay against the same hint.
    static let regionToleranceSeconds: TimeInterval = 0.05

    /// pickLowestRTT returns the relay with the smallest measured
    /// RTT, or nil when:
    ///   * the candidate list is empty
    ///   * every probe failed (caller logs the cause and degrades
    ///     to direct-only; not an error since the network may
    ///     have direct-only peers anyway)
    ///
    /// preferredRegion is the §4 P2 stage 5 affinity hint. Empty ⇒
    /// pure RTT ranking (back-compat with pre-stage-5 behaviour).
    /// When set, a same-region relay wins over a lower-RTT out-of-
    /// region one if the gap is within `regionToleranceSeconds`.
    ///
    /// Parallel by design: serial would stretch a 5-relay probe to
    /// 5s in the worst case, doubling tunnel bring-up latency the
    /// user can see.
    static func pickLowestRTT(
        from candidates: [BambooClient.RelayServer],
        preferredRegion: String = ""
    ) async -> BambooClient.RelayServer? {
        guard !candidates.isEmpty else { return nil }

        // TaskGroup fans out probes in parallel and collects
        // (index, RTT-or-error) pairs. We retain the original
        // index so the rank-sort can still resolve back to the
        // RelayServer struct without copying it into the group.
        let results: [(Int, TimeInterval)] = await withTaskGroup(
            of: (Int, TimeInterval?).self
        ) { group in
            for (idx, rs) in candidates.enumerated() {
                group.addTask {
                    let outcome = await probeFn(rs.hostname, rs.port)
                    switch outcome {
                    case .success(let rtt):
                        return (idx, rtt)
                    case .failure:
                        return (idx, nil)
                    }
                }
            }
            var ok: [(Int, TimeInterval)] = []
            for await (idx, rtt) in group {
                if let rtt = rtt {
                    ok.append((idx, rtt))
                }
            }
            return ok
        }
        guard !results.isEmpty else { return nil }
        let sorted = results.sorted { $0.1 < $1.1 }
        let winnerIdx = applyRegionAffinity(
            candidates: candidates,
            ranked: sorted,
            preferredRegion: preferredRegion
        )
        return candidates[winnerIdx]
    }

    /// applyRegionAffinity returns the index of the winning relay
    /// after the §4 P2 stage 5 same-region preference. Empty
    /// preferredRegion ⇒ RTT winner. If the RTT winner is already
    /// in the preferred region, no-op. Otherwise the first same-
    /// region row within `regionToleranceSeconds` of the RTT
    /// winner wins; if none qualifies, the RTT winner stays. The
    /// hint can therefore never make the picker pathologically
    /// slow — only same or better than pure RTT.
    private static func applyRegionAffinity(
        candidates: [BambooClient.RelayServer],
        ranked: [(Int, TimeInterval)],
        preferredRegion: String
    ) -> Int {
        let rttWinnerIdx = ranked[0].0
        if preferredRegion.isEmpty { return rttWinnerIdx }
        if candidates[rttWinnerIdx].region == preferredRegion { return rttWinnerIdx }
        let winnerRTT = ranked[0].1
        for (idx, rtt) in ranked {
            guard candidates[idx].region == preferredRegion else { continue }
            if rtt - winnerRTT <= regionToleranceSeconds {
                return idx
            }
            // ranked is ASC by RTT; once a same-region row exceeds
            // the tolerance, all later same-region rows do too.
            break
        }
        return rttWinnerIdx
    }

    /// relayWSSURL builds the wss URL for a chosen relay. Wraps
    /// the (hostname, port) → URL conversion in one place so a
    /// future scheme change (h3, quic) lives here rather than
    /// being threaded through every dial site.
    static func relayWSSURL(_ rs: BambooClient.RelayServer) -> URL? {
        return URL(string: "wss://\(rs.hostname):\(rs.port)/relay")
    }
}

/// probeRelayRTT measures TCP-connect latency to one (hostname,
/// port). Uses NWConnection — Apple's modern Network framework
/// API — which works on macOS 10.14+ / iOS 12+ without falling
/// back to BSD sockets.
///
/// Returns .success(rtt) when the connection reaches .ready
/// state, .failure(timeoutOrErr) when the probe deadline trips
/// or the underlying connection errors.
///
/// The serial dispatch queue is what makes single-resume safe:
/// stateUpdateHandler + asyncAfter(timeout) both run on it, so
/// the resumed flag is accessed without any cross-thread races.
internal func probeRelayRTT(hostname: String, port: Int) async -> Result<TimeInterval, Error> {
    let host = NWEndpoint.Host(hostname)
    // UInt16(exactly:) returns nil on overflow rather than trapping
    // — a future caller passing a parsed-from-JSON port that
    // exceeds 65535 must surface as a probe failure, not a crash.
    guard let port16 = UInt16(exactly: port),
          let portNum = NWEndpoint.Port(rawValue: port16) else {
        return .failure(RelayPickerError.invalidPort(port))
    }
    let conn = NWConnection(host: host, port: portNum, using: .tcp)
    // Per-probe serial queue. Hand-rolled label aids os_log
    // attribution if a future deadlock investigation needs to
    // grep for which probe queue is stuck.
    let queue = DispatchQueue(label: "dev.hanfour.bamboo.relay-probe.\(hostname)")
    let start = Date()

    // ProbeGate wraps the single-resume flag in a reference type
    // so closures captured by the serial queue can share state
    // without tripping Swift 6's sendable-capture rules. All
    // mutation happens on `queue`, which is serial — the
    // @unchecked Sendable conformance reflects that.
    final class ProbeGate: @unchecked Sendable {
        var resumed = false
    }
    let gate = ProbeGate()

    return await withCheckedContinuation { cont in
        queue.asyncAfter(deadline: .now() + RelayPicker.probeTimeout) {
            if !gate.resumed {
                gate.resumed = true
                conn.cancel()
                cont.resume(returning: .failure(RelayPickerError.timeout))
            }
        }
        conn.stateUpdateHandler = { state in
            switch state {
            case .ready:
                if !gate.resumed {
                    gate.resumed = true
                    let rtt = Date().timeIntervalSince(start)
                    conn.cancel()
                    cont.resume(returning: .success(rtt))
                }
            case .failed(let err):
                if !gate.resumed {
                    gate.resumed = true
                    conn.cancel()
                    cont.resume(returning: .failure(err))
                }
            case .waiting(let err):
                // NWConnection enters .waiting when the path is
                // temporarily unavailable (cellular off, transient
                // DNS fail). For a fast-bring-up probe we treat
                // this as a failure rather than waiting indefinitely
                // — the timeout would catch it anyway, but failing
                // fast lets a healthy relay win sooner.
                if !gate.resumed {
                    gate.resumed = true
                    conn.cancel()
                    cont.resume(returning: .failure(err))
                }
            default:
                break
            }
        }
        conn.start(queue: queue)
    }
}

internal enum RelayPickerError: Error, CustomStringConvertible {
    case timeout
    case invalidPort(Int)

    public var description: String {
        switch self {
        case .timeout:                return "RelayPicker: probe timed out"
        case .invalidPort(let p):     return "RelayPicker: invalid port \(p)"
        }
    }
}
